systems/controllers/cp_traj_gen.cc

```diff
@@ -0,0 +1,397 @@
+                BasicVector<double>(1)).get_index();
+  if (is_using_predicted_com) {
+    com_port_ = this->DeclareAbstractInputPort("CoM_traj",
+        drake::Value<ExponentialPlusPiecewisePolynomial<double>> {}).get_index();
+  }
+  if (is_walking_position_control) {
+    fp_port_ = this->DeclareVectorInputPort(BasicVector<double>(2)).get_index();
+  }
+
+  this->DeclareAbstractOutputPort("cp_traj", &CPTrajGenerator::CalcTrajs);
+
+  // State variables inside this controller block
+  DeclarePerStepDiscreteUpdateEvent(&CPTrajGenerator::DiscreteVariableUpdate);
+  // The swing foot position at touchdown
+  prev_td_swing_foot_idx_ = this->DeclareDiscreteState(3);
+  // The time of the last touch down
+  prev_td_time_idx_ = this->DeclareDiscreteState(1);
+  // The last state of FSM
+  prev_fsm_state_idx_ = this->DeclareDiscreteState(-0.1 * VectorXd::Ones(1));
+
+  // Check if the model is floating based
+  is_quaternion_ = multibody::CheckFloatingBase(tree);
+}
+
+
+EventStatus CPTrajGenerator::DiscreteVariableUpdate(
+  const Context<double>& context,
+  DiscreteValues<double>* discrete_state) const {
+
+  // Read in finite state machine
+  const BasicVector<double>* fsm_output = (BasicVector<double>*)
+      this->EvalVectorInput(context, fsm_port_);
+  VectorXd fsm_state = fsm_output->get_value();
+
+  auto prev_fsm_state = discrete_state->get_mutable_vector(
+                          prev_fsm_state_idx_).get_mutable_value();
+
+  if (fsm_state(0) != prev_fsm_state(0)) { //if at touchdown
+    prev_fsm_state(0) = fsm_state(0);
+
+    auto swing_foot_pos_td = discrete_state->get_mutable_vector(
+                               prev_td_swing_foot_idx_).get_mutable_value();
+    auto prev_td_time = discrete_state->get_mutable_vector(
+                          prev_td_time_idx_).get_mutable_value();
+
+    // Read in current state
+    const OutputVector<double>* robot_output = (OutputVector<double>*)
+        this->EvalVectorInput(context, state_port_);
+
+    // Get time
+    double timestamp = robot_output->get_timestamp();
+    double current_time = static_cast<double>(timestamp);
+    prev_td_time(0) = current_time;
+
+    // Kinematics cache and indices
+    KinematicsCache<double> cache = tree_->CreateKinematicsCache();
+    VectorXd q = robot_output->GetPositions();
+    // Modify the quaternion in the begining when the state is not received from
+    // the robot yet (cannot have 0-norm quaternion when using doKinematics)
+    if (is_quaternion_) {
+      q.segment(3, 4) = NormalizeQuaternion(q.segment(3, 4));
+    }
+    cache.initialize(q);
+    tree_->doKinematics(cache);
+    int swing_foot_idx = (fsm_state(0) == right_stance_) ?
+                           left_foot_idx_ : right_foot_idx_;
+    Vector3d pt_on_swing_foot = (fsm_state(0) == right_stance_) ?
+                                pt_on_left_foot_ : pt_on_right_foot_;
+
+    // Swing foot position (Forward Kinematics) and velocity at touchdown
```

Comment:
This forward kinematics can be done via a single Drake call, you should almost never have to manually deal with rotation matrices. This should do the trick (where the `0` is the world index):

```
swing_foot_pos_td = tree_->transformPoints(cache, pt_on_swing_foot, swing_foot_idx, 0);
```